filterDemo/katcam/katcam.go

```diff
@@ -0,0 +1,183 @@
+	VisionService string `json:"visionService"`
+}
+
+func (cfg *Config) Validate(path string) ([]string, error) {
+	if cfg.ActualCam == "" {
+		return nil, fmt.Errorf(`expected "actualCam" attribute for katcam %q`, path)
+	}
+
+	return []string{cfg.ActualCam}, nil
+}
+
+type katCam struct {
+	resource.Named
+	actualCam camera.Camera
+	logger    golog.Logger
+
+	// // Case 1: Only store one out of every 10 captures.
+	// count *int32
+
+	// Case 2:
+	visionService vision.Service
+}
+
+// resource.Resource methods
+func (c *katCam) Reconfigure(ctx context.Context, deps resource.Dependencies, conf resource.Config) error {
+	c.actualCam = nil
+	camConfig, err := resource.NativeConfig[*Config](conf)
+	if err != nil {
+		return err
+	}
+
+	c.actualCam, err = camera.FromDependencies(deps, camConfig.ActualCam)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get camera %v for katcam", camConfig.ActualCam)
+	}
+
+	// Case 2:
+	c.visionService, err = vision.FromDependencies(deps, camConfig.VisionService)
+	if err != nil {
+		return errors.Wrapf(err, "unable to get vision service %v for katcam", camConfig.VisionService)
+	}
+
+	return nil
+}
+
+func (c *katCam) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
+	return cmd, nil
+}
+
+func (c *katCam) Close(ctx context.Context) error {
+	return c.actualCam.Close(ctx)
+}
+
+// VideoStream methods
+func (c *katCam) Images(ctx context.Context) ([]image.Image, time.Time, error) {
+	return nil, time.Time{}, errUnimplemented
+}
+
+// Can implement Read instead of Stream - investigate which cams are MediaReader vs. MediaStream?
+// func (c *katCam) Read(ctx context.Context) (image.Image, func(), error) {
+// 	return nil, nil, nil
+// }
+
+func (c *katCam) Stream(ctx context.Context, errHandlers ...gostream.ErrorHandler) (gostream.VideoStream, error) {
+	camStream, err := c.actualCam.Stream(ctx, errHandlers...)
+	if err != nil {
+		return nil, err
+	}
+	// filterStream := filterStream{camStream}
+
+	// // Case 1:
+	// filterStream := filterStream{camStream, c.count}
+
+	// Case 2:
+	filterStream := filterStream{camStream, c.visionService}
+
+	return filterStream, nil
+}
+
+func (c *katCam) NextPointCloud(ctx context.Context) (pointcloud.PointCloud, error) {
+	return nil, errUnimplemented
+}
+
+func (c *katCam) Properties(ctx context.Context) (camera.Properties, error) {
+	return camera.Properties{}, nil
+}
+
+// Projector methods
+func (c *katCam) Projector(ctx context.Context) (transform.Projector, error) {
+	return nil, errUnimplemented
+}
+
+// Filter code:
+type filterStream struct {
+	cameraStream gostream.VideoStream
+
+	// // Case 1: Only store one out of every 10 captures.
+	// count *int32
+
+	// Case 2:
+	visionService vision.Service
+}
+
+func (fs filterStream) Next(ctx context.Context) (image.Image, func(), error) {
+	if ctx.Value(data.CtxKeyDM) != true {
+		return nil, nil, errors.New("Cannot access filter stream if not DM collector")
```

Comment:
If it's not a data management collector, we want to return the underlying camera stream.